cli/fossilize_replay.cpp

```diff
@@ -2568,28 +2550,53 @@ struct ThreadedReplayer : StateCreatorInterface
 			work_item.memory_context_index = PARENT_PIPELINE_MEMORY_CONTEXT;
 			work_item.tag = DerivedInfo::get_tag();
 			enqueue_work_item(work_item);
+
+			parsed_parents.insert(h);
 		}
 	}
 
 	template <typename DerivedInfo>
 	void enqueue_parent_pipelines_pipeline_library(const DerivedInfo &info,
-												   const VkPipelineLibraryCreateInfoKHR &library_info,
-	                                               const unordered_map<Hash, VkPipeline> &pipelines)
+	                                               const VkPipelineLibraryCreateInfoKHR &library_info,
+	                                               const unordered_map<Hash, VkPipeline> &pipelines,
+	                                               unordered_map<Hash, DerivedInfo> &parents,
+	                                               unordered_set<Hash> &parsed_parents)
 	{
 		for (uint32_t i = 0; i < library_info.libraryCount; i++)
 		{
 			Hash h = (Hash)library_info.pLibraries[i];
-			enqueue_parent_pipeline(info, h, pipelines);
+			enqueue_parent_pipeline(info, h, pipelines, parents, parsed_parents);
 		}
 	}
 
 	template <typename DerivedInfo>
 	void enqueue_parent_pipelines(const DerivedInfo &info,
-	                              const unordered_map<Hash, VkPipeline> &pipelines)
+	                              const unordered_map<Hash, VkPipeline> &pipelines,
+	                              unordered_map<Hash, DerivedInfo> &parents,
+	                              unordered_set<Hash> &parsed_parents)
 	{
 		auto *library = work_item_get_library_info(info);
 		if (library)
-			enqueue_parent_pipelines_pipeline_library(info, *library, pipelines);
+			enqueue_parent_pipelines_pipeline_library(info, *library, pipelines, parents, parsed_parents);
+	}
+
+	template <typename DerivedInfo>
+	void compute_parents_depth(const unordered_map<Hash, DerivedInfo> &parents,
+	                           unordered_map<Hash, uint32_t> &parents_depth,
+	                           const DerivedInfo &info, uint32_t depth)
+	{
+		parents_depth[info.hash] = max(parents_depth.at(info.hash), depth);
+
+		auto library = work_item_get_library_info(info);
```

Comment:
This leads to exponential explosion as well. I think what can be done here is to compare depth against parents depth. Only if depth is greater than parents_depth does the recursive descend happen. That way we can terminate the traversal early.